docs/adr/0002-recipe-rewrite.md

```diff
@@ -0,0 +1,254 @@
+
+`RecipeInput`s are a list of `RecipeInputItem`s plus a `MatchProcedure` -- how the inputs of
+the recipe should be matched to items in a multiblock/machine when crafting. The base ones are:
+
+- Shaped/Shapeless: Exactly the same as vanilla
+- Subset: How the current smeltery, etc. craft
+- Shaped-flippable: The recipe can be flipped on the Y-axis
+- Shaped-rotatable: The recipe can be rotated (currently only 45deg, 3x3)
+
+`RecipeInputItem`s describe a single slot of a recipe and determines what
+items match it. There can be a single item that matches (see `RecipeInputSlimefunItem`,
+`RecipeInputItemStack`), or a list (tag) of items all of which can be used
+in that slot (see `RecipeInputGroup`, `RecipeInputTag`).
+
+`RecipeOutput`s are just a list of `RecipeOutputItem`s, all of which are crafted by the recipe.
+
+An `RecipeOutputItem`s controls how an output is generated when the recipe is
+crafted. It can be a single item (see `RecipeOutputItemStack`, `RecipeOutputSlimefunItem`),
+or a group of items each with a certain weight of being output (see `RecipeOutputGroup`).
+
+#### Examples (pseudocode)
+
+Here are the inputs and outputs of the recipe for a vanilla torch
+
+```txt
+RecipeInput (
+    {
+        EMPTY, EMPTY, EMPTY
+        EMPTY, RecipeInputGroup(COAL, CHARCOAL), EMPTY,
+        EMPTY, RecipeInputItemStack(STICK), EMPTY
+    },
+    SHAPED
+)
+RecipeOutput (
+    RecipeOutputItemStack(4 x TORCH)
+)
+```
+
+Here are the inputs and outputs of a gold pan
+
+```txt
+RecipeInput (
+    { RecipeOutputItemStack(GRAVEL) },
+    SUBSET
+)
+RecipeOutput (
+    RecipeOutputGroup(
+        40 RecipeOutputItemStack(FLINT)
+        5  RecipeOutputItemStack(IRON_NUGGET)
+        20 RecipeOutputItemStack(CLAY_BALL)
+        35 RecipeOutputSlimefunItem(SIFTED_ORE)
+    )
+)
+```
+
+This would remove the need to use ItemSettings to determine the gold pan weights
+
+### RecipeService
+
+This is the public interface for the recipe system, there are methods here to add,
+load, save, and search recipes. It also stores a map of `MatchProcedures` and
+`RecipeType` by key for conversions from a string
+
+### JSON Serialization
+
+All recipes are able to be serialized to and deserialized
+from JSON. The schemas are shown below.
+
+Here, `key` is the string representation of a namespaced key
+
+`Recipe`
+
+```txt
+{
+    "input"?: RecipeInput
+    "output"?: RecipeOutput
+    "type": key | key[]
+    "energy"?: int
+    "crafting-time"?: int
+    "permission-node"?: string | string[]
+}
+```
+
+The recipe deserializer technically needs a `__filename` field, but it is
+inserted when the file is read, so it isn't (and shouldn't) be in the schema
+
+`RecipeInput`
+
+```txt
+{
+    "items": string | string[]
+    "key": {
+        [key: string]: RecipeInputItem
+    }
+    "match"?: key
+}
+```
+
+`RecipeOutput`
+
+```txt
+{
+    "items": RecipeOutputItem[]
+}
+```
+
+`RecipeInputItem`*
+
+```txt
+{
+    "id": key
+    "amount"?: int
+    "durability"?: int
```

Comment:
why durability?

Reply:
i remember some mods that had recipes like this, where you could put an item and some kind of tool in, and when you crafted, it would use up some dura on the tool (i *think* it was farmers delight but i could be misremembering)

i thought that would be cool to add, but if you think its out of scope or should just be a custom recipe ill remove it